File: difusioIC.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
using Subset = vector<int>;

// Struct for undirected graph
struct Graph {
    int numNodes;
    vector<vector<int> > adjList;

    // Graph constructor:
    Graph(int n) {
        numNodes = n;
        adjList.resize(n);
    }

    Graph() {
    }

    void addEdge(int u, int v) {
        adjList[u].push_back(v);
        adjList[v].push_back(u);
    }

    void printEdges() {
        for (int vertex = 0; vertex < adjList.size(); ++vertex) {
            for (int neighbour : adjList[vertex]) cout << vertex << " -> " << neighbour << endl;
        }
    }
};

// Simulate IC difusion process
int simulateIC(Graph& G, double& p, Subset& S, int& t) {
    // Set of influenced nodes
    vector<bool> influenced(G.numNodes, false);
    // Set of active nodes (per iteration)
    queue<int> active;
    
    t = 0;

    // Activate initial set of nodes
    for (int vertex : S) {
        influenced[vertex] = true;
        active.push(vertex);
    }

    // Continue untill no more active nodes
    while (not active.empty()) {
        ++t;
        int numCurrentActive = active.size();
        // Process all active nodes in the current layer
        for (int i = 0; i < numCurrentActive; ++i) {
            int v = active.front();
            active.pop();
            // Check neighbours of v
            for (int j = 0; j < G.adjList[v].size(); ++j) {
                int neighbour = G.adjList[v][j];
                 // If neighbor is not already active, try to activate it
                if (not influenced[neighbour]) {
                    double r = (double) rand() / RAND_MAX;
                    if (r < p) {
                        influenced[neighbour] = true;
                        active.push(neighbour);
                    }
                }
            }
        }
    }

    int count = 0;
    for (int i = 0; i < influenced.size(); ++i) if (influenced[i]) ++count; 
    return count;
}

void readInput(Graph& G, double& p, Subset& S) {
    int n, m;
    cout << "Enter number of nodes and number of edges: ";
    cin >> n >> m;
    G = Graph(n);
    cout << "Enter " << m << " edges: ";
    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        G.addEdge(u, v);
    }
    cout << "Enter probability for IC model: ";
    cin >> p;
    int s;
    cout << "Enter number of nodes in the initial seed: ";
    cin >> s;
    cout << "Enter " << s << " nodes for the initial seed: ";
    for (int i = 0; i < s; ++i) {
        int u;
        cin >> u;
        S.push_back(u);
    }
}

int main() {

    // Generate random seed for proper random values
    unsigned seed = chrono::high_resolution_clock::now().time_since_epoch().count();
    srand(seed);

    Graph G;
    double p;
    Subset S;
    readInput(G, p, S);
    int t = 0;
    int C = simulateIC(G, p, S, t);
    cout << "Size of C: " << C << endl;
    cout << "Value of t: " << t << endl;
}

File: difusioIC.py
import warnings
warnings.filterwarnings("ignore")

import random
import networkx as nx
import matplotlib.pyplot as plt

# Struct for undirected graph
class Graph:
    def __init__(self, n):
        self.numNodes = n
        self.adjList = [[] for i in range(n)]
        
    def addEdge(self, u, v):
        self.adjList[u].append(v)
        self.adjList[v].append(u)

# Simulate IC diffusion process
def simulateIC(G, p, S, nx_G, pos, node_colors):
    # Set of influenced nodes
    influenced = [False] * G.numNodes
    # Set of active nodes (per iteration)
    active = []
    
    # Number of steps
    t = 0

    # Order of nodes as stored in the graph, lo necesitamos
    node_order = list(nx_G.nodes())

    # Activate initial set of nodes
    for vertex in S:
        influenced[vertex] = True
        active.append(vertex)

    # Correct "order" for node color list, if we don't do this the nodes are painted wrong
    i = 0
    for node in node_order:
        node_colors[i] = 'r' if influenced[node] else 'grey'
        i += 1

    # Paint the graph for the first time
    nx.draw_networkx_nodes(nx_G, pos, nodelist=node_order, node_color=node_colors, cmap=plt.cm.Reds)
    plt.draw()
    plt.pause(1)

    # Continue until no more active nodes
    while active:
        t += 1
        numCurrentActive = len(active)
        # Process all active nodes in the current layer
        for i in range(numCurrentActive):
            v = active.pop(0)
            # Check neighbours of v
            for neighbour in G.adjList[v]:
                # If neighbor is not already active, try to activate it
                if not influenced[neighbour]:
                    r = random.random() # random number between 0 and 1
                    if r < p:
                        influenced[neighbour] = True
                        active.append(neighbour)

        # Checking again if we need to color new nodes
        # Correct "order" for node color list, if we don't do this the nodes are painted wrong
        i = 0
        for node in node_order:
            node_colors[i] = 'r' if influenced[node] else 'grey'
            i += 1

        nx.draw_networkx_nodes(nx_G, pos, nodelist=node_order, node_color=node_colors, cmap=plt.cm.Reds)
        plt.draw()
        plt.pause(1)

    count = 0
    for node in influenced:
        if node:
            count += 1
    return count, t, influenced

def readInput():
    n, m = map(int, input("Enter number of nodes and number of edges: ").split())
    G = Graph(n)
    print(f"Enter {m} edges:")
    for i in range(m):
        u, v = map(int, input().split())
        G.addEdge(u, v)
    p = float(input("Enter probability for IC model: "))
    s = int(input("Enter number of nodes in the initial seed: "))
    print(f"Enter {s} nodes for the initial seed:")
    S = list(map(int, input().split()))
    return G, p, S

if __name__ == "__main__":
    G, p, S = readInput()

    # Convert adjacency list to list of edges
    edge_list = []
    for u in range(G.numNodes):
        for v in G.adjList[u]:
            if (u, v) not in edge_list and (v, u) not in edge_list:
                edge_list.append((u, v))

    # Create NetworkX graph from the edge list
    nx_G = nx.Graph()
    nx_G.add_edges_from(edge_list)

    # Draw the graph
    pos = nx.spring_layout(nx_G)

    node_colors = ['grey' for _ in range(nx_G.number_of_nodes())]
    nx.draw(nx_G, pos, with_labels=True)

    # Run simulation and update the node colors
    C, t, influenced = simulateIC(G, p, S, nx_G, pos, node_colors)

    print("Size of C:", C)
    print("Value of t:", t)

    # Show the final plot
    plt.show(block=True)
